### PR TITLE
Refactor (and fix) nosp/sms classification logic

### PR DESCRIPTION
**Pull request by kjdchapman**
This PR attempts to make the business rules around 'escaping' a particular classification, e.g.  
> if it's got an agreement on, don't send a NOSP.

This refactoring/exploration exposed some weird conditions/edge cases. This PR includes fixing the behaviour for these cases.

Notably, there is still duplication across each classification, the tests are hard to read, and other classification methods are still opaque. I hope to address these next.

### CODE REVIEW

File: spec/lib/hackney/income/tenancy_prioritiser/classifications/send_nosp_spec.rb [hunk 1]
```diff
@@ -29,17 +29,6 @@
       eviction_date: nil,
       courtdate: ''
     },
-    # nosp has been served, other data isn't right, no other detail
-    # not clear why this test is passing when `@criteria.nosp_served? == false &&` is removed
```
Comment (kjdchapman) [inline spec/lib/hackney/income/tenancy_prioritiser/classifications/send_nosp_spec.rb]: This is now clear to me - it's because of the (removed in this commit!) default in `stub_criteria.rb`.

File: lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb [hunk 2]
```diff
@@ -64,12 +64,13 @@ def apply_for_court_date?
...
         def send_sms?
-          @criteria.last_communication_action.nil? &&
-            @criteria.balance >= 5 &&
-            @criteria.nosp_served? == false &&
-            last_communication_between_three_months_one_week? &&
-            @case_priority.paused? == false &&
-            @criteria.active_agreement? == false
+          return false if @criteria.last_communication_action.present?
+          return false if @criteria.nosp_served?
+          return false unless last_communication_between_three_months_one_week?
```
[the remark this answers is not in the document]
Comment (maysakanoni) [reply lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]: I think this may be because you only dealing with the nosps in this pr?

[the remark this answers is not in the document]
Comment (kjdchapman) [reply lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]: Yeah, I didn't deliberately make any actual changes to SMS behaviour in this PR. I think once this is all sorted it's a lot easier to see what the rules "should" be!

That said, I think you're right - this logic looks like it'll send SMSs too aggressively to me, and I think I should look at it as a piece soon.

[the remark this answers is not in the document]
Comment (richrace) [reply lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]: It won't send more than one SMS due to line 67.

[the remark this answers is not in the document]
Comment (kjdchapman) [reply lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]: Good spot. It won't send SMSs aggressively _enough_, then!
